compiler/rustc_mir_build/src/thir/pattern/mod.rs

```diff
@@ -575,63 +574,39 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
             }
         };
 
-        let cid = GlobalId { instance, promoted: None };
-        // Prefer valtrees over opaque constants.
-        let const_value = self
-            .tcx
-            .const_eval_global_id_for_typeck(param_env_reveal_all, cid, span)
-            .map(|val| match val {
-                Some(valtree) => mir::Const::Ty(ty, ty::Const::new_value(self.tcx, valtree, ty)),
-                None => mir::Const::Val(
-                    self.tcx
-                        .const_eval_global_id(param_env_reveal_all, cid, span)
-                        .expect("const_eval_global_id_for_typeck should have already failed"),
-                    ty,
-                ),
-            });
-
-        match const_value {
-            Ok(const_) => {
-                let pattern = self.const_to_pat(const_, id, span);
-
-                if !is_associated_const {
-                    return pattern;
-                }
+        let c = ty::Const::new_unevaluated(
+            self.tcx,
+            ty::UnevaluatedConst { def: instance.def_id(), args: instance.args },
+        );
```

Comment:
This code looks a bit odd... we create an instance only to immediately take it apart again. Can this be simplified further? We have `def_id` and `args` already before the `let instance = ...`, can we just use those directly? But there are some specific error cases up there that I don't know how to handle then, in particular `AssocConstInPattern`.

Reply:
If I remove the `Instance::try_resolve`, then the only change in the tests is that "associated consts cannot be referenced in patterns" becomes "constant pattern depends on a generic parameter". Which seems okay? It *does* depend on a generic parameter:
```rust
pub fn test<A: Foo, B: Foo>(arg: EFoo, A::X: EFoo) {
    //~^ ERROR associated consts cannot be referenced in patterns
    let A::X = arg;
    //~^ ERROR associated consts cannot be referenced in patterns
}
```
If the associated const can be resolved, I am fairly sure we already accept that code, so the general claim "associated consts cannot be referenced in patterns" is already not true I think.

Any objections to changing the error message for that case?

Reply:
I think that change is fine :+1: